### PR TITLE
Create a wrapper around Kord for interacting with Discord

### LINKED ISSUE #7: Discord integration
We need to be able to connect to Discord. Preferably, this would be a wrapper so that any third-party dependency is contained to one module.

### PR DESCRIPTION
**Pull request by boswelja**
This PR sets up the basics for our first bot in terms of Discord integration. We have some builder interfaces that support the bare minimum needed for Dale to function.

Closes #7

### CODE REVIEW

File: dale/src/main/kotlin/Main.kt [hunk 1]
```diff
@@ -16,7 +16,7 @@
 import guildsettings.guildSettingDatabaseInstance
 import studio.AndroidStudioUpdateChecker
 
-suspend fun main(args: Array<String>) {
+suspend fun main(args: Array<String>) = discordBot(args.first()) {
```
Comment (boswelja) [inline dale/src/main/kotlin/Main.kt:19]: We'll need to think about proper arg handling next I think

File: core/discord/src/main/kotlin/kord/interaction/KordApplicationCommandScope.kt [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.flow.filterIsInstance
+import kotlinx.coroutines.launch
+
+internal class KordApplicationCommandScope(
+    private val restClient: RestClient,
+    private val gateway: Gateway,
+    private val scope: CoroutineScope,
+) : ApplicationCommandScope {
+    override suspend fun registerGlobalChatInputCommand(
```
Comment (danielesegato) [inline core/discord/src/main/kotlin/kord/interaction/KordApplicationCommandScope.kt:34]: I don't think this needs to be a suspend function

File: core/discord/src/main/kotlin/kord/presence/KordPresenceScope.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                    this.status = when (status) {
+                        PresenceStatus.INVISIBLE -> dev.kord.common.entity.PresenceStatus.Invisible
+                        PresenceStatus.DO_NOT_DISTURB -> dev.kord.common.entity.PresenceStatus.DoNotDisturb
+                        PresenceStatus.IDLE -> dev.kord.common.entity.PresenceStatus.Idle
+                        PresenceStatus.ONLINE -> dev.kord.common.entity.PresenceStatus.Online
+                    }
+                }
+            }
+        }
+    }
```
Comment (danielesegato) [inline core/discord/src/main/kotlin/kord/presence/KordPresenceScope.kt:45]: why execute it as a side-effect coroutine instead of a suspend function?

**Review by danielesegato — COMMENTED** [review]
while I generally like the work you've done here I'm a bit puzzled by the usage of coroutines.

There are a lot of functions that spawn side effects and I'd like to understand a bit better the rationale behind them

File: core/discord/src/main/kotlin/DiscordBot.kt [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        callsInPlace(block, InvocationKind.EXACTLY_ONCE)
+    }
+
+    supervisorScope {
+        val bot = createKordDiscordBot(token)
+
+        launch {
+            block(bot)
+        }
+    }
```
[the remark this answers is not in the document]
Comment (boswelja) [reply core/discord/src/main/kotlin/DiscordBot.kt:53]: I wanted a scope that wouldn't take out the entire application if something went wrong, if there's a better way please let me know

File: core/discord/src/main/kotlin/kord/interaction/KordApplicationCommandScope.kt (hunk 2 above)
Comment (boswelja) [reply core/discord/src/main/kotlin/kord/interaction/KordApplicationCommandScope.kt:34]: Registering the command makes a web request, which is suspending as per Kord's API.

I do like that the way it's currently set up will suspend until the command is registered, but I'm happy to make it non-suspending and just `launch` the whole thing if you'd rather

File: core/discord/src/main/kotlin/kord/presence/KordPresenceScope.kt (hunk 3 above)
Comment (boswelja) [reply core/discord/src/main/kotlin/kord/presence/KordPresenceScope.kt:45]: I can't see on mobile but I'm guessing this is for the `launch` that collects the SharedFlow.

I don't think it makes sense to suspend execution indefinitely when we register a command, which would require us to `launch` the register call every time for a functional bot.

Comment (danielesegato) [reply core/discord/src/main/kotlin/kord/presence/KordPresenceScope.kt:45]:
```kotlin
class KordPresenceScope(
    private val token: String,
    private val gateway: Gateway,
    private val scope: CoroutineScope,
) : PresenceScope {
    override fun updatePresence(afk: Boolean, status: PresenceStatus) {
        scope.launch {
            gateway.stop()
            gateway.start(token) {
                presence {
                    this.afk = afk
                    this.status = when (status) {
                        PresenceStatus.INVISIBLE -> dev.kord.common.entity.PresenceStatus.Invisible
                        PresenceStatus.DO_NOT_DISTURB -> dev.kord.common.entity.PresenceStatus.DoNotDisturb
                        PresenceStatus.IDLE -> dev.kord.common.entity.PresenceStatus.Idle
                        PresenceStatus.ONLINE -> dev.kord.common.entity.PresenceStatus.Online
                    }
                }
            }
        }
    }
```
the issue I have with this is that it breaks multiple boundaries
start/stop gateway and side effect

when i'm in a situation like this I prefer to have a single component in charge of receiving changes and another that can push them to it in a queue of some sort.

so you could have a single coroutine listening to this queue and performing changes

I don't want to block development over this tho'

File: core/discord/src/main/kotlin/kord/interaction/KordApplicationCommandScope.kt (hunk 2 above)
Comment (danielesegato) [reply core/discord/src/main/kotlin/kord/interaction/KordApplicationCommandScope.kt:34]: those are 2 completely different things tho'

one is listening to commands and publish them to observer, the other is the network call.

Unless I'm missing something

**Pull request merged**